### PR TITLE
Send email to milestone recipient after every collect/disburse

### PR DESCRIPTION
**Pull request by mohammadranjbarz**
Now we send email after every user collects/disburse donation payments 

implement #246

### CODE REVIEW

File: src/services/milestones/sendNotification.js [hunk 1]
```diff
@@ -241,15 +241,15 @@ const sendNotification = () => async context => {
...
-        Notifications.milestonePaid(app, {
-          recipient: recipient.email,
-          user: recipient.name,
-          milestoneTitle: title,
-          milestoneId: _id,
-          donationCounters,
-          address: recipientAddress,
-          campaignId,
-        });
+        // Notifications.milestonePaid(app, {
```
Comment (mohammadranjbarz) [inline src/services/milestones/sendNotification.js:244]: @aminlatifi 
I think these are should be delete but I commented it to know your opinion for that.

File: src/utils/dappMailer.js [hunk 2]
```diff
@@ -410,4 +410,32 @@ module.exports = {
...
+        <p>You can expect to see these payment(s) to arrive in your wallet <em>${
+          data.address
+        }</em> within 48 - 72 hrs.</p>
+      `,
+      cta: `See your Milestones`,
+      ctaRelativeUrl: `/campaigns/${data.campaignId}/milestones/${data.milestoneId}`,
+      unsubscribeType: 'donations-collected',
+      unsubscribeReason: `You receive this email because you are the recipient of a Milestone`,
+    });
+    sendEmail(app, data);
```
Comment (mohammadranjbarz) [inline src/utils/dappMailer.js:439]: @aminlatifi 
I wrote subjects and texts by myself, if you think there are some changes needed for texts let me know

File: src/services/donations/updateEntityCounters.js [hunk 3]
```diff
@@ -147,75 +147,79 @@ const updateEntity = async (app, id, type) => {
...
-                { performedByAddress: donation.actionTakerAddress },
-              );
-            }
-          } catch (e) {
-            logger.error('could not create conversation', e);
-          } finally {
-            conversationSem.leave();
-          }
+      // why our linter has problem with for..of ?
+      /* eslint-disable no-restricted-syntax */
```
Comment (mohammadranjbarz) [inline src/services/donations/updateEntityCounters.js:186]: @aminlatifi 

Why our linter has problem with for..of,  is there a logical reason or it's just a legacy linter?

File: src/services/donations/updateEntityCounters.js [hunk 4]
```diff
@@ -147,75 +148,80 @@ const updateEntity = async (app, id, type) => {
...
-              const conversation = data[0];
-              const payments = conversation.payments || [];
-              const index = payments.findIndex(p => p.symbol === symbol);
+    const { txHash } = donation;
+    const events = (
+      await app.service('events').find({
+        query: {
+          transactionHash: donation.txHash,
+          event: 'Transfer',
+          status: { $nin: [EventStatus.PROCESSED, EventStatus.FAILED] },
```
Comment (aminlatifi) [inline src/services/donations/updateEntityCounters.js:161]: Did you test it? the event become processed after event is handled. So when the last event is handled, corresponding event would have processing state.... so the result length will not be zero at all and no email is sent!

Comment (mohammadranjbarz) [reply src/services/donations/updateEntityCounters.js:161]: Yes I tested locally,  although for doesn't have accessing to email server the email not sent successfully but I printed the log, and the conversation created ( for one donation and two donation of different tokens I tested)

event became Processed and after that this function calls, so when reaching to this line if this event is the last one be sure that this query will return zero events

Comment (mohammadranjbarz) [reply src/services/donations/updateEntityCounters.js:161]: @aminlatifi 

The reason why Donation was created after event status became `Processed` is not putting await before an async function, your guess was right,  Do you think we need to fix that problem during this issue (I vote for yes)?
https://github.com/Giveth/feathers-giveth/blob/d64f82c383a133c96c8b69874e7da06140725409/src/blockchain/pledges.js#L460-L465

**Comment by mohammadranjbarz** [comment]
@aminlatifi 

I fixed to change event status to `Processed` after donation is created,
And now support multi donations with similar token but different old pledges, these screenshots is from my testing in local env

image


image

**Comment by mohammadranjbarz** [comment]
I couldn't find out why the mechanism of counting unProcessed transfer events doesnt work, but through inspecting for that I found anothr interesting thing
The last event that processed when a payment collected ends is `PaymentCollected` so we can move the creating conversation for payment and email to milestone recipient in handling PaymentCollected event handler It's more readable for the developers who will contribute to codes in future, When I checked for every transactionHash I saw the paymentCollected is the last event that processed( and updated) you can check that too:

https://feathers.beta.giveth.io/events?transactionHash=0xb26da0f569413c20a73c1ad60e377b9f5edf68f503f8d9f86553149d0ef138ba&$sort[updatedAt]=1&$select[]=event&$select[]=updatedAt

https://feathers.beta.giveth.io/events?transactionHash=0x984631bad86b71e31c00f5d35875ad44749eb8e4ea145fb35554e55c84839081&$sort[updatedAt]=1&$select[]=event&$select[]=updatedAt

https://feathers.beta.giveth.io/events?transactionHash=0x0d65ad429fff6eab905d9f2c4f09066e81013fe75faf11f0c08c0394bd8425af&$sort[updatedAt]=1&$select[]=event&$select[]=updatedAt

**Comment by mohammadranjbarz** [comment]
@aminlatifi  
It tested and  I think it's better now, handling in `PaymentCollected` is more straight-forward and easy to understand

File: src/blockchain/milestones.js [hunk 5]
```diff
@@ -248,23 +335,29 @@ const milestonesFactory = app => {
         );
         return;
       }
-
-      // if (!milestone.maxAmount || !milestone.fullyFunded) return;
-      // never set uncapped or non-fullyFunded milestones as PAID
-      if (!matchingMilestones[0].maxAmount || !matchingMilestones[0].fullyFunded) return;
```
Comment (aminlatifi) [inline src/blockchain/milestones.js:254]: This part should not be removed

File: src/services/milestones/sendNotification.js (hunk 1 above)
Comment (aminlatifi) [reply src/services/milestones/sendNotification.js:244]: Yes remove it

Comment (aminlatifi) [reply src/services/milestones/sendNotification.js:244]: Also remove milestonePaid function definition.

File: src/blockchain/milestones.js (hunk 5 above)
Comment (mohammadranjbarz) [reply src/blockchain/milestones.js:254]: It wasn't removed already, it just moved to some lines later.


image

File: src/services/milestones/sendNotification.js (hunk 1 above)
Comment (mohammadranjbarz) [reply src/services/milestones/sendNotification.js]: > Also remove milestonePaid function definition.

Done.

**Comment by mohammadranjbarz** [comment]
@horus2021 
Could you please test it on develop  env?

User should donate/delegate a milestone, and after every collect/disburse a conversation should be create including all donations amount, and after the the email should send to milestone's recipient (no matter if milestone is capped or not)

**Pull request merged**